cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/GetRevisionTaskThreaded.java

```diff
@@ -0,0 +1,210 @@
+            @Override
+            public boolean add(Future<DocumentRevsList> documentRevsListFuture) {
+                boolean offerResult;
+                try {
+                    offerResult = offer(documentRevsListFuture, 5, TimeUnit.MINUTES);
+                } catch (InterruptedException ie) {
+                    throw new RuntimeException("Offer interrupted", ie);
+                }
+                if (!offerResult) {
+                    throw new RuntimeException("Offer timed out");
+                }
+                return true;
+            }
+        };
+        this.completionService = new ExecutorCompletionService<DocumentRevsList>(executorService, responses);
+        this.requestsOutstanding = new AtomicInteger(requests.size());
+
+        // we make the request at construction time...
+        for (final BulkGetRequest request : requests) {
+
+            completionService.submit(new Callable<DocumentRevsList>() {
+                @Override
+                public DocumentRevsList call() throws Exception {
+                    return new DocumentRevsList(GetRevisionTaskThreaded.this.sourceDb.getRevisions
+                            (request.id, request.revs, request.atts_since,
+                                    GetRevisionTaskThreaded.this
+                                            .pullAttachmentsInline));
+                }
+            });
+        }
+    }
+
+    @Override
+    public Iterator<DocumentRevsList> iterator() {
+        return new GetRevisionTaskIterator();
+    }
+
+    @Override
+    public String toString() {
+        return "GetRevisionTask{" +
+                "sourceDb=" + sourceDb +
+                ", requests=" + requests +
+                ", pullAttachmentsInline=" + pullAttachmentsInline +
+                '}';
+    }
+
+    private class GetRevisionTaskIterator implements Iterator<DocumentRevsList> {
+
+        @Override
+        public boolean hasNext() {
+            // can't advance iterator as there was a problem in call()
+            if (!iteratorValid) {
+                return false;
+            }
+
+            boolean next = requestsOutstanding.get() != 0;
+
+            // if we have returned all of our results, we can shut down the executor
+            if (!next) {
+                iteratorValid = false;
+                executorService.shutdown();
+                try {
+                    executorService.awaitTermination(5, TimeUnit.SECONDS);
+                } catch (InterruptedException e) {
+                    ;
+                }
+            }
+
+            return next;
+        }
+
+        @Override
+        public DocumentRevsList next() {
+            // can't advance iterator as there was a problem in call()
+            if (!iteratorValid) {
+                return null;
+            }
+
+            try {
+                requestsOutstanding.decrementAndGet();
+                return completionService.take().get();
```

Comment:
Do you think we should use `completionService.poll(5, TimeUnit.MINUTES).get()` to match up with the `offer`? Otherwise I'm concerned that this iterator could block forever (say if the network vanishes) unless we have another timeout somewhere that guarantees the `InterruptedException` will get thrown.